core/src/main/java/org/apache/calcite/rel/rules/ProjectTableScanRule.java

```diff
@@ -103,30 +106,47 @@ protected static boolean test(TableScan scan) {
   protected void apply(RelOptRuleCall call, Project project, TableScan scan) {
     final RelOptTable table = scan.getTable();
     assert table.unwrap(ProjectableFilterableTable.class) != null;
-    if (!project.isMapping()) {
-      return;
-    }
-    final Mappings.TargetMapping mapping = Project.getPartialMapping(
-            project.getInput().getRowType().getFieldCount(),
-            project.getProjects());
 
-    final ImmutableIntList projects;
-    final ImmutableList<RexNode> filters;
+    final List<Integer> selectedColumns = new ArrayList<>();
+    project.getProjects().forEach(proj -> {
+      proj.accept(new RexVisitorImpl<Void>(true) {
+        public Void visitInputRef(RexInputRef inputRef) {
+          if (!selectedColumns.contains(inputRef.getIndex())) {
+            selectedColumns.add(inputRef.getIndex());
+          }
+          return null;
+        }
+      });
+    });
```

Comment:
Can we use InputFinder here? And it is better to use bit set instead of int list.

Reply:
Well, I was also thinking about this ~
The concern here is that the order of selected fields will be lost if using `InputFinder` and record by bit set.
If the order of selected fields for pushing down is lost, we have to add another `Project` on top of `BindableTableScan` for adjusting. 
`ScannableTableTest#testPushSimpleMappingProject` is to test such case.